File: frontend/src/services/formatters.js
/**
 * Centralized missing-data and presentation formatting policy for PackCheck frontend.
 *
 * Strictly enforces:
 * 1. Missing scalar values -> "Not available in recorded evidence"
 * 2. Missing extracted declarations -> Value: "Not available in recorded evidence", Status: "Not recorded"
 * 3. Missing confidence -> "Not available in recorded evidence" (never 0 or 1)
 * 4. Missing source region -> "Not available in recorded evidence" (never estimated coordinates)
 * 5. Missing quality assessment -> "Not available in recorded evidence"
 * 6. Missing rule fields -> "Not available in recorded evidence"
 * 7. Empty review actions -> "No human review actions recorded."
 * 8. Empty collections:
 *    - Evidence -> "No evidence records available."
 *    - Observations -> "No observations recorded."
 *    - Rule evaluations -> "No rule evaluations recorded."
 *    - Review actions -> "No human review actions recorded."
 * 9. Explicitly not-applicable data -> "Not applicable"
 * 10. Canonical dispositions preserved: PASS, FAIL, REVIEW_REQUIRED, NOT_ASSESSABLE
 *
 * Never renders raw null, undefined, None, blank strings, or N/A.
 */

export const NOT_AVAILABLE = 'Not available in recorded evidence';
export const NOT_RECORDED = 'Not recorded';
export const NOT_APPLICABLE = 'Not applicable';

export const EMPTY_EVIDENCE = 'No evidence records available.';
export const EMPTY_OBSERVATIONS = 'No observations recorded.';
export const EMPTY_RULES = 'No rule evaluations recorded.';
export const EMPTY_REVIEWS = 'No human review actions recorded.';
export const EMPTY_HISTORY = 'No inspections recorded.';

export const MANDATORY_DECLARATIONS = [
  { key: 'product_name', label: 'Product Name' },
  { key: 'manufacturer_name', label: 'Manufacturer Name' },
  { key: 'manufacturer_address', label: 'Manufacturer Address' },
  { key: 'packer_name', label: 'Packer Name' },
  { key: 'importer_name', label: 'Importer Name' },
  { key: 'net_quantity', label: 'Net Quantity' },
  { key: 'mrp', label: 'MRP' },
  { key: 'month_year_of_manufacture', label: 'Manufacturing Date (Month/Year)' },
  { key: 'consumer_care_details', label: 'Consumer Care Details' },
];

export const FIELD_LABELS = MANDATORY_DECLARATIONS.reduce((acc, curr) => {
  acc[curr.key] = curr.label;
  return acc;
}, {});

/**
 * Format any scalar value according to the missing-data policy.
 */
export function formatMissingValue(value, isApplicable = true, fallback = NOT_AVAILABLE) {
  if (!isApplicable) return NOT_APPLICABLE;
  if (value === null || value === undefined) return fallback;

  const s = String(value).trim();
  if (!s || ['none', 'null', 'undefined', 'n/a'].includes(s.toLowerCase())) {
    return fallback;
  }

  if (['not_applicable', 'not applicable', 'n.a.', 'na'].includes(s.toLowerCase())) {
    return NOT_APPLICABLE;
  }

  return s;
}

/**
 * Format an extraction confidence score.
 */
export function formatConfidence(confidence, isApplicable = true) {
  if (!isApplicable) return NOT_APPLICABLE;
  if (confidence === null || confidence === undefined) return NOT_AVAILABLE;

  if (typeof confidence === 'string') {
    const s = confidence.trim();
    if (!s || ['none', 'null', 'undefined', 'n/a'].includes(s.toLowerCase())) {
      return NOT_AVAILABLE;
    }
    if (s === NOT_AVAILABLE || s === NOT_APPLICABLE) return s;
    const parsed = parseFloat(s);
    if (!isNaN(parsed)) {
      return `${(parsed * 100).toFixed(0)}%`;
    }
    return s;
  }

  const val = parseFloat(confidence);
  if (isNaN(val)) return NOT_AVAILABLE;
  return `${(val * 100).toFixed(0)}%`;
}

/**
 * Format bounding box coordinates into string or fallback.
 */
export function formatSourceRegion(sourceRegion, isApplicable = true) {
  if (!isApplicable) return NOT_APPLICABLE;
  if (!sourceRegion) return NOT_AVAILABLE;

  if (typeof sourceRegion === 'string') {
    const s = sourceRegion.trim();
    if (!s || ['none', 'null', 'undefined', 'n/a'].includes(s.toLowerCase())) {
      return NOT_AVAILABLE;
    }
    return s;
  }

  const { x, y, width, height } = sourceRegion;
  if (x === undefined || y === undefined || width === undefined || height === undefined) {
    return NOT_AVAILABLE;
  }

  return `[x:${x}, y:${y}, ${width}×${height}px]`;
}

/**
 * Format quality gate assessment summary.
 */
export function formatQualitySummary(quality) {
  if (!quality) return NOT_AVAILABLE;
  const isAcceptable = quality.is_acceptable;
  const score = quality.overall_score !== undefined ? Number(quality.overall_score).toFixed(2) : null;
  const statusStr = isAcceptable ? 'PASS (Acceptable)' : 'REJECTED (Insufficient)';
  return score !== null ? `${statusStr} — Score: ${score} / 1.00` : statusStr;
}

/**
 * Get CSS theme styling classes/colors for canonical compliance dispositions.
 */
export function getDispositionTheme(disposition) {
  const d = String(disposition || '').toUpperCase();
  switch (d) {
    case 'PASS':
      return {
        badgeClass: 'badge-pass',
        bannerClass: 'banner-pass',
        label: 'PASS',
        color: 'var(--accent-emerald)',
        icon: '✓',
      };
    case 'FAIL':
      return {
        badgeClass: 'badge-fail',
        bannerClass: 'banner-fail',
        label: 'FAIL',
        color: 'var(--accent-red)',
        icon: '✗',
      };
    case 'REVIEW_REQUIRED':
      return {
        badgeClass: 'badge-review',
        bannerClass: 'banner-review',
        label: 'REVIEW REQUIRED',
        color: 'var(--accent-amber)',
        icon: '⚠',
      };
    case 'NOT_ASSESSABLE':
    default:
      return {
        badgeClass: 'badge-not-assessable',
        bannerClass: 'banner-not-assessable',
        label: 'NOT ASSESSABLE',
        color: 'var(--text-muted)',
        icon: '○',
      };
  }
}

File: backend/app/services/formatters.py
"""Centralized missing-data and display policy formatters.

Enforces a single missing-data rendering policy across the Result API and PDF Report:
1. Missing / null scalar values -> "Not available in recorded evidence"
2. Missing extracted declaration -> Value: "Not available in recorded evidence", Status: "Not recorded"
3. Missing confidence -> "Not available in recorded evidence" (never assumed 0 or 1)
4. Missing source region -> "Not available in recorded evidence" (never estimated coordinates)
5. Missing quality assessment -> "Not available in recorded evidence"
6. Missing rule fields -> "Not available in recorded evidence"
7. Missing human review -> "No human review actions recorded."
8. Empty collections:
   - Evidence -> "No evidence records available."
   - Observations -> "No observations recorded."
   - Rule evaluations -> "No rule evaluations recorded."
   - Review actions -> "No human review actions recorded."
9. Explicitly not-applicable data -> "Not applicable"
10. Preserves canonical statuses (PASS, FAIL, REVIEW_REQUIRED, NOT_ASSESSABLE).
"""

from typing import Any, List, Optional

# Canonical display constants
NOT_AVAILABLE_MESSAGE = "Not available in recorded evidence"
NOT_APPLICABLE_MESSAGE = "Not applicable"
NOT_RECORDED_MESSAGE = "Not recorded"

EMPTY_EVIDENCE_MESSAGE = "No evidence records available."
EMPTY_OBSERVATIONS_MESSAGE = "No observations recorded."
EMPTY_RULE_EVALUATIONS_MESSAGE = "No rule evaluations recorded."
EMPTY_REVIEW_ACTIONS_MESSAGE = "No human review actions recorded."

# Expected standard packaging declaration fields
MANDATORY_DECLARATION_FIELDS: List[str] = [
    "product_name",
    "manufacturer_name",
    "manufacturer_address",
    "packer_name",
    "importer_name",
    "net_quantity",
    "mrp",
    "month_year_of_manufacture",
    "consumer_care_details",
]


def format_missing_value(
    value: Optional[Any],
    is_applicable: bool = True,
    fallback: str = NOT_AVAILABLE_MESSAGE,
) -> str:
    """Format a scalar value according to the missing-data policy.

    - Returns 'Not applicable' if is_applicable is False or value explicitly indicates N/A.
    - Returns 'Not available in recorded evidence' if value is None, empty, or blank string.
    - Never returns raw None, null, undefined, blank strings, or N/A.
    - Preserves canonical compliance verdicts (PASS, FAIL, REVIEW_REQUIRED, NOT_ASSESSABLE).
    """
    if not is_applicable:
        return NOT_APPLICABLE_MESSAGE

    if value is None:
        return fallback

    s = str(value).strip()
    if not s or s.lower() in ("none", "null", "undefined", "n/a"):
        return fallback

    if s.lower() in ("not_applicable", "not applicable", "n.a.", "na"):
        return NOT_APPLICABLE_MESSAGE

    return s


def format_confidence(confidence: Optional[Any], is_applicable: bool = True) -> str:
    """Format an extraction confidence score.

    - If unavailable, returns 'Not available in recorded evidence'.
    - Never converts missing confidence to 0, 1, or an assumed value.
    """
    if not is_applicable:
        return NOT_APPLICABLE_MESSAGE

    if confidence is None:
        return NOT_AVAILABLE_MESSAGE

    if isinstance(confidence, str):
        s = confidence.strip()
        if not s or s.lower() in ("none", "null", "undefined", "n/a"):
            return NOT_AVAILABLE_MESSAGE
        if s.lower() in ("not_applicable", "not applicable"):
            return NOT_APPLICABLE_MESSAGE
        if s == NOT_AVAILABLE_MESSAGE:
            return NOT_AVAILABLE_MESSAGE
        try:
            val = float(s)
            return f"{val:.2f}"
        except ValueError:
            return s

    try:
        val = float(confidence)
        return f"{val:.2f}"
    except (ValueError, TypeError):
        return NOT_AVAILABLE_MESSAGE


def format_source_region(source_region: Optional[Any], is_applicable: bool = True) -> str:
    """Format source bounding box region.

    - If unavailable, returns 'Not available in recorded evidence'.
    - Never creates or estimates coordinates.
    """
    if not is_applicable:
        return NOT_APPLICABLE_MESSAGE

    if source_region is None:
        return NOT_AVAILABLE_MESSAGE

    if isinstance(source_region, str):
        s = source_region.strip()
        if not s or s.lower() in ("none", "null", "undefined", "n/a"):
            return NOT_AVAILABLE_MESSAGE
        if s.lower() in ("not_applicable", "not applicable"):
            return NOT_APPLICABLE_MESSAGE
        return s

    if isinstance(source_region, dict):
        if not source_region:
            return NOT_AVAILABLE_MESSAGE
        x = source_region.get("x")
        y = source_region.get("y")
        w = source_region.get("width")
        h = source_region.get("height")
        if x is None or y is None or w is None or h is None:
            return NOT_AVAILABLE_MESSAGE
        return f"[{x}, {y}, {w}x{h}]"

    if hasattr(source_region, "x") and hasattr(source_region, "y"):
        return f"[{source_region.x}, {source_region.y}, {source_region.width}x{source_region.height}]"

    s = str(source_region).strip()
    return s if s else NOT_AVAILABLE_MESSAGE


def format_quality_summary(quality: Optional[Any]) -> str:
    """Format quality gate assessment summary."""
    if quality is None:
        return NOT_AVAILABLE_MESSAGE

    # QualityAssessment instance
    if hasattr(quality, "overall_score") and hasattr(quality, "is_acceptable"):
        status_str = "PASS (Acceptable)" if quality.is_acceptable else "REJECTED (Insufficient)"
        return f"{status_str} — Score: {quality.overall_score:.2f} / 1.00"

    if isinstance(quality, dict):
        if not quality or "overall_score" not in quality:
            return NOT_AVAILABLE_MESSAGE
        is_acc = quality.get("is_acceptable", False)
        status_str = "PASS (Acceptable)" if is_acc else "REJECTED (Insufficient)"
        return f"{status_str} — Score: {float(quality['overall_score']):.2f} / 1.00"

    if isinstance(quality, str):
        s = quality.strip()
        if not s or s.lower() in ("none", "null", "undefined", "n/a"):
            return NOT_AVAILABLE_MESSAGE
        return s

    return NOT_AVAILABLE_MESSAGE


def clean_presentation_dict(data: Any) -> Any:
    """Recursively sanitize presentation dictionaries to ensure no raw null/None/undefined/N/A."""
    if isinstance(data, dict):
        cleaned = {}
        for k, v in data.items():
            if v is None:
                if k == "evidence":
                    cleaned[k] = NOT_AVAILABLE_MESSAGE
                elif k == "quality":
                    cleaned[k] = NOT_AVAILABLE_MESSAGE
                elif k in ("source_region", "source_region_before", "source_region_after"):
                    cleaned[k] = NOT_AVAILABLE_MESSAGE
                elif k in ("confidence", "original_confidence"):
                    cleaned[k] = NOT_AVAILABLE_MESSAGE
                elif k == "review_actions":
                    cleaned[k] = []
                elif k == "observations":
                    cleaned[k] = []
                elif k == "rule_evaluations":
                    cleaned[k] = []
                else:
                    cleaned[k] = NOT_AVAILABLE_MESSAGE
            elif isinstance(v, str):
                cleaned[k] = format_missing_value(v)
            elif isinstance(v, (dict, list)):
                cleaned[k] = clean_presentation_dict(v)
            else:
                cleaned[k] = v
        return cleaned
    elif isinstance(data, list):
        return [clean_presentation_dict(item) for item in data]
    elif data is None:
        return NOT_AVAILABLE_MESSAGE
    elif isinstance(data, str):
        return format_missing_value(data)
    return data
